ipld/unixfs/io/directory.go

```diff
@@ -554,49 +758,58 @@ func (d *DynamicDirectory) AddChild(ctx context.Context, name string, nd ipld.No
 		}
 
 		if switchToBasic {
-			basicDir, err := hamtDir.switchToBasic(ctx)
+			basicDir, err := hamtDir.switchToBasic(ctx, WithMaxLinks(hamtDir.maxLinks), WithCidBuilder(hamtDir.GetCidBuilder()))
 			if err != nil {
 				return err
 			}
 			err = basicDir.AddChild(ctx, name, nd)
 			if err != nil {
 				return err
 			}
-			d.Directory = basicDir
+			d.directoryWithOptions = basicDir
 			return nil
 		}
 
-		return d.Directory.AddChild(ctx, name, nd)
+		return d.directoryWithOptions.AddChild(ctx, name, nd)
 	}
 
 	// BasicDirectory
-	basicDir := d.Directory.(*BasicDirectory)
+	basicDir := d.directoryWithOptions.(*BasicDirectory)
 	switchToHAMT, err := basicDir.needsToSwitchToHAMTDir(name, nd)
 	if err != nil {
 		return err
 	}
 	if !switchToHAMT {
 		return basicDir.AddChild(ctx, name, nd)
 	}
-	hamtDir, err = basicDir.switchToSharding(ctx)
+
+	maxLinks := DefaultShardWidth
+	// Verify that our maxLinks is usuable for ShardWidth (power of 2, multiple of 8)
+	if validShardWidth(basicDir.maxLinks) {
+		maxLinks = basicDir.maxLinks
+	} else if basicDir.maxLinks > 0 {
+		log.Warning("using DefaultShardWidth for HAMT directories. MaxLinks cannot be used as it is not a power of 2 and multiple of 8.")
```

Comment:
I think `log.Warn` is what is normally used and `log.Warning` may actually be deprecated, but I am not sure. 
```suggestion
		log.Warn("using DefaultShardWidth for HAMT directories. MaxLinks cannot be used as it is not a power of 2 and multiple of 8.")
```